### PR TITLE
Adding news app

### LINKED ISSUE #3: Create app for news
Depends on #2.
This app would be used to manage news on the website.

### Features
* Extends announcements
* Ability to add display start time.
* Content with HTML support.
* News body with HTML support.
* Views
  * Paginated list at `/news/`
  * Single view at `/news/<id>/`

Also, show latest 3 (should be configurable) news at index.

### PR DESCRIPTION
**Pull request by Monal5031**
Fixed #3

### CODE REVIEW

File: wsgi/odyssy/basic/templates/basic/index.html [hunk 1]
```diff
@@ -1,3 +1,12 @@
 <body>
     Welcome to IIIT Vadodara.
-</body>
\ No newline at end of file
+    {% if latest_news %}
+        <ul>
+            {% for news in latest_news %}
+                <li><a href="/news/{{ news.id }}"> {{ news.title }}</a></li>
```
Comment (Prakash2403) [inline wsgi/odyssy/basic/templates/basic/index.html]: As mentioned in django documentation, using `{% url %}` tag is preferred.

File: wsgi/odyssy/news/models.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    """
+    start_date = models.DateTimeField()
+    end_date = models.DateTimeField()
+    title = models.CharField(max_length=50)
+    description = models.CharField(max_length=500)
+
+    def __str__(self):
+        return str(self.pk) + ': ' + str(self.title)
+
+    def save(self, force_insert=False, force_update=False, using=None, update_fields=None):
```
Comment (Prakash2403) [inline wsgi/odyssy/news/models.py]: This method is fine, but it has it's own disadvantages. Refer to [second answer at this link](https://stackoverflow.com/questions/2281179/adding-extra-constraints-into-fields-in-django)

File: wsgi/odyssy/news/templates/news/news_list.html [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    {% if news_list %}
+        <ul>
+            {% for news in news_list %}
+                <li><a href="/news/{{ news.pk }}">{{ news.title }}</a></li>
+            {% endfor %}
+        </ul>
+    {% else %}
+        <h4>No News to display</h4>
+    {% endif %}
+</body>
```
Comment (Prakash2403) [inline wsgi/odyssy/news/templates/news/news_list.html]: Add pagination here.

File: wsgi/odyssy/news/views.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
+from django.shortcuts import render
+from .models import News
+
+
+# Create your views here.
+def open_all_news(request):
+    news = News.objects.order_by('-start_date')
+    paginator = Paginator(news, 10)
```
Comment (Prakash2403) [inline wsgi/odyssy/news/views.py]: Embed number of items in `GET` request. Extract it from there.

File: wsgi/odyssy/news/urls.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from django.conf.urls import url
+
+from . import views
+
+urlpatterns = [
```
Comment (Monal5031) [inline wsgi/odyssy/news/urls.py]: Any idea over this issue?

Comment (yashLadha) [inline wsgi/odyssy/news/urls.py]: try to add a pylint flag
` # pylint: disable=invalid-name `
Hope this helps

File: wsgi/odyssy/news/urls.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from django.conf.urls import url
+
+from . import views
+
+urlpatterns = [  # pylint: disable=invalid-name
```
Comment (singhpratyush) [inline wsgi/odyssy/news/urls.py]: Please remove the comment as this project doesn't need to be consistent with `Pylint`. It is there just for pointing out and we can ignore the things that we wish to.

File: wsgi/odyssy/news/views.py [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+        news = news.get()
+    else:
+        news = None
+    context = {
+        'info': news,
+    }
+    return render(request, 'news/single_news.html', context)
+
+
+def latest_news(num_items):
```
Comment (singhpratyush) [inline wsgi/odyssy/news/views.py]: Add this and all queried that you use as `staticmethod` to respective class.

File: wsgi/odyssy/odyssy/settings.py [hunk 8]
```diff
@@ -57,6 +57,7 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'news',
```
Comment (singhpratyush) [inline wsgi/odyssy/odyssy/settings.py]: Why not `news.apps.NewsConfig`?

File: wsgi/odyssy/odyssy/urls.py [hunk 9]
```diff
@@ -18,6 +18,7 @@
 
 urlpatterns = [
     url(r'^admin/?', include(admin.site.urls)),
-    url(r'^events/', include('events.urls')),
     url(r'^', include('basic.urls')),
+    url(r'^news/', include('news.urls')),
+    url(r'^events/', include('events.urls')),
```
Comment (singhpratyush) [inline wsgi/odyssy/odyssy/urls.py]: Please do not touch the code which is irrelevant to this pull request.

File: wsgi/odyssy/news/admin.py [hunk 10]
```diff
@@ -0,0 +1,5 @@
+from django.contrib import admin
+from .models import News
+# Register your models here.
+
+admin.site.register(News)
```
Comment (singhpratyush) [inline wsgi/odyssy/news/admin.py]: Two lines break before.

File: wsgi/odyssy/news/urls.py (hunk 6 above)
Comment (singhpratyush) [inline wsgi/odyssy/news/urls.py]: Two line break before.

File: wsgi/odyssy/odyssy/urls.py (hunk 9 above)
Comment (Monal5031) [inline wsgi/odyssy/odyssy/urls.py]: I was resolving the conflicts, I don't know how this is showing as my addition. 😕

File: wsgi/odyssy/odyssy/settings.py (hunk 8 above)
Comment (Monal5031) [inline wsgi/odyssy/odyssy/settings.py]: Is there any particular difference?
or are we just following that pattern?

File: wsgi/odyssy/news/urls.py (hunk 5 above)
Comment (Monal5031) [inline wsgi/odyssy/news/urls.py]: Thanks that worked. 😄

File: wsgi/odyssy/odyssy/urls.py (hunk 9 above)
Comment (singhpratyush) [inline wsgi/odyssy/odyssy/urls.py]: Please include `basic` app at last.

**Comment by Monal5031** [comment]
@singhpratyush Please review again.

File: wsgi/odyssy/news/views.py [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+        news = news.get()
+    else:
+        news = None
+    context = {
+        'info': news,
+    }
+    return render(request, 'news/single_news.html', context)
+
+
+@staticmethod
```
Comment (singhpratyush) [inline wsgi/odyssy/news/views.py]: This is not the change that I was requesting. You need to add all the methods you require to class itself.

File: wsgi/odyssy/odyssy/settings.py [hunk 12]
```diff
@@ -57,6 +57,7 @@
     'django.contrib.sessions',
     'django.contrib.messages',
     'django.contrib.staticfiles',
+    'news.apps.NewsConfig',
```
Comment (singhpratyush) [inline wsgi/odyssy/odyssy/settings.py]: Let's keep custom apps on the top.

File: wsgi/odyssy/news/views.py [hunk 13]
```diff
@@ -0,0 +1,41 @@
+from django.core.paginator import Paginator, EmptyPage
+from django.shortcuts import render
+from .models import News
+
+
+# Create your views here.
+
+@staticmethod
+def open_all_news(request):
```
Comment (singhpratyush) [inline wsgi/odyssy/news/views.py]: Please make function names consistent with events app.

File: wsgi/odyssy/basic/views.py [hunk 14]
```diff
@@ -1,8 +1,15 @@
 from django.shortcuts import render
+from news.models import News
 
 from events.models import Event
 
 
 def index(request):
+    """ Index page of the Website """
```
Comment (singhpratyush) [inline wsgi/odyssy/basic/views.py]: Inappropriate comment style.

File: wsgi/odyssy/news/models.py [hunk 15]
```diff
@@ -0,0 +1,41 @@
+import datetime
+
+from django.core.exceptions import ValidationError
+from django.db import models
+from django.shortcuts import get_object_or_404
+
+# Create your models here.
```
Comment (singhpratyush) [inline wsgi/odyssy/news/models.py]: Remove this comment.

File: wsgi/odyssy/news/models.py [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+
+
+class News(models.Model):
+    """ Model of News app
```
Comment (singhpratyush) [inline wsgi/odyssy/news/models.py]: Inappropriate comment style.

File: wsgi/odyssy/news/models.py [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+    start_date - Start Date of News
+    end_date - End Date of News
+    title - Title for News
+    description - Description of News
+    """
+    start_date = models.DateTimeField(default=datetime.datetime.now())
```
Comment (singhpratyush) [inline wsgi/odyssy/news/models.py]: Shouldn't it be
```python
start_date = models.DateTimeField(default=datetime.datetime.now)
```
or even
```python
start_date = models.DateTimeField(default=timezone.now)
```

File: wsgi/odyssy/news/templates/news/single_news.html [hunk 18]
```diff
@@ -0,0 +1,8 @@
+<body>
+    {% if info %}
+        <h1>{{ info.title }}</h1>
+        <p>{{ info.description }}</p>
+    {% else %}
+        <h1>No Such News present</h1>
+    {% endif %}
+</body>
```
Comment (singhpratyush) [inline wsgi/odyssy/news/templates/news/single_news.html]: Add newline.

File: wsgi/odyssy/news/models.py (hunk 15 above)
Comment (Monal5031) [inline wsgi/odyssy/news/models.py]: Done.

File: wsgi/odyssy/news/models.py (hunk 16 above)
Comment (Monal5031) [inline wsgi/odyssy/news/models.py]: Done, made changes according to https://www.python.org/dev/peps/pep-0257/#multi-line-docstrings

File: wsgi/odyssy/news/templates/news/single_news.html (hunk 18 above)
Comment (Monal5031) [inline wsgi/odyssy/news/templates/news/single_news.html]: Done.

File: wsgi/odyssy/basic/views.py (hunk 14 above)
Comment (Monal5031) [inline wsgi/odyssy/basic/views.py]: Done.

File: wsgi/odyssy/news/models.py (hunk 17 above)
Comment (Monal5031) [inline wsgi/odyssy/news/models.py]: True, the second is much better and also no need of parentheses. 👍

Comment (singhpratyush) [inline wsgi/odyssy/news/models.py]: Why did you find the second one better? Please share a link describing it.

File: wsgi/odyssy/news/models.py (hunk 16 above)
Comment (singhpratyush) [inline wsgi/odyssy/news/models.py]: The description should be in the second line
```python
class News(models.Model):
    """
    Model of News app
    """
     ...
```

File: wsgi/odyssy/news/models.py (hunk 17 above)
Comment (Monal5031) [inline wsgi/odyssy/news/models.py]: https://stackoverflow.com/questions/10783864/django-1-4-timezone-now-vs-datetime-datetime-now.

**Pull request merged**